src/java/org/apache/cassandra/index/sai/plan/QueryViewBuilder.java

```diff
@@ -18,92 +18,147 @@
     {
-        this.expressions = expressions;
+        this.cfs = cfs;
+        this.orderer = orderer;
         this.range = range;
+        this.queryContext = queryContext;
     }
 
-    public static class QueryView
+    public static class QueryView implements AutoCloseable
     {
-        public final Map<SSTableReader, List<IndexExpression>> view;
-        public final Set<SSTableIndex> referencedIndexes;
-
-        public QueryView(Map<SSTableReader, List<IndexExpression>> view, Set<SSTableIndex> referencedIndexes)
+        final ColumnFamilyStore.RefViewFragment view;
+        final Set<SSTableIndex> referencedIndexes;
+        final Set<MemtableIndex> memtableIndexes;
+        final Orderer orderer;
+
+        public QueryView(ColumnFamilyStore.RefViewFragment view,
+                         Set<SSTableIndex> referencedIndexes,
+                         Set<MemtableIndex> memtableIndexes,
+                         Orderer orderer)
         {
             this.view = view;
             this.referencedIndexes = referencedIndexes;
+            this.memtableIndexes = memtableIndexes;
+            this.orderer = orderer;
+        }
+
+        @Override
+        public void close()
+        {
+            view.release();
+            referencedIndexes.forEach(SSTableIndex::release);
         }
     }
 
     /**
-     * Acquire references to all the SSTableIndexes required to query the expressions.
+     * Acquire references to all the memtables, memtable indexes, sstables, and sstable indexes required for the
+     * given expression.
      * <p>
      * Will retry if the active sstables change concurrently.
      */
     protected QueryView build()
     {
-        Set<SSTableIndex> referencedIndexes = new HashSet<>();
-        AtomicBoolean failed = new AtomicBoolean();
+        var referencedIndexes = new HashSet<SSTableIndex>();
+        long failingSince = -1L;
         try
         {
+            outer:
             while (true)
             {
-                referencedIndexes.clear();
-                failed.set(false);
-
-                Map<SSTableReader, List<IndexExpression>> view = getQueryView(expressions);
-                view.values()
-                .stream()
-                .flatMap(expressions -> expressions.stream().map(e -> e.index))
-                .forEach(index ->
+                // Prevent an infinite loop
+                queryContext.checkpoint();
+
+                // Acquire live memtable index and memtable references first to avoid missing an sstable due to flush.
+                // Copy the memtable indexes to avoid concurrent modification.
+                var memtableIndexes = new HashSet<>(orderer.context.getLiveMemtables().values());
+
+                // TODO live sstables or canonical?
+                var refViewFragment = cfs.selectAndReference(org.apache.cassandra.db.lifecycle.View.selectLive(range));
```

Comment:
Does anyone know whether we should prefer live or canonical sstables here?

Reply:
I fixed this in this commit https://github.com/datastax/cassandra/pull/1073/commits/0bdd4801445d53eeabc97f6c8d4a3d209a0eec72. We'll use canonical sstables.